### PR TITLE
Add dialyzer target

### PR DESCRIPTION
**Pull request by ppikula**
The idea of this PR is to add specs to all amoc's functions and testing in on Travis.

### CODE REVIEW

**Comment by studzien** [comment]
Looks like ``types.hrl`` is missing :(

File: src/amoc_controller.erl [hunk 1]
```diff
@@ -43,41 +48,55 @@
...
 %% ------------------------------------------------------------------
 %% gen_server Function Definitions
 %% ------------------------------------------------------------------
+-spec init([]) -> {ok, state()}.
 init([]) ->
     process_flag(priority, max),
     State = #state{scenario = undefined},
     {ok, State}.
 
+-spec handle_call(any(), any(), state()) -> {reply, handle_call_resp(), state()}.
```
Comment (gadgetci) [inline src/amoc_controller.erl]: According to **Elvis**:
> Line 99 is too long: -spec handle_call(any(), any(), state()) -> {reply, handle_call_resp(), state()}..

File: src/amoc_slave.erl [hunk 2]
```diff
@@ -109,13 +126,14 @@ ping_slave_nodes(#state{to_ack=Ack}=State) ->
     end,
     State#state{to_ack=Ack1}.
 
+-spec ping_slave_node({node(), no_retries()}) -> false | {true, {node(), no_retries()}}.
```
Comment (gadgetci) [inline src/amoc_slave.erl]: According to **Elvis**:
> Line 129 is too long: -spec ping_slave_node({node(), no_retries()}) -> false | {true, {node(), no_retries()}}..

File: src/amoc_annotations.erl [hunk 3]
```diff
@@ -62,7 +75,10 @@ annotate(Tags, Format, Args) ->
             ok
     end.
 
-annotation({dist_do, Scenario, Start, End, Nodes, Comment}) ->
+-spec annotation(command()) -> {tag(), io:format(), list()}  | ignore.
+annotation({dist_do, Scenario, Start, End, Opts}) ->
+    Nodes = proplists:get_value(nodes, Opts, nodes()),
+    Comment = proplists:get_value(comment, Opts, "none"),
     {<<"amoc start">>, "Scenario: ~p. comment: ~p Start: ~p. End: ~p. Nodes: ~p.",
```
Comment (gadgetci) [inline src/amoc_annotations.erl]: According to **Elvis**:
> Line 82 is too long:     {<<"amoc start">>, "Scenario: ~p. comment: ~p Start: ~p. End: ~p. Nodes: ~p.",.

File: src/amoc_slave.erl [hunk 4]
```diff
@@ -41,22 +47,26 @@ ping(Node) ->
...
 
 %% ------------------------------------------------------------------
 %% gen_server Function Definitions
 %% ------------------------------------------------------------------
+-spec init([]) -> {ok, state()}.
 init([]) ->
     schedule_timer(),
     {ok, #state{to_ack = []}}.
 
+-spec handle_call(command(), {pid(), any()}, state()) -> {reply, ok | pong, state()}.
```
Comment (gadgetci) [inline src/amoc_slave.erl]: According to **Elvis**:
> Line 66 is too long: -spec handle_call(command(), {pid(), any()}, state()) -> {reply, ok | pong, state()}..

File: src/amoc_controller.erl [hunk 5]
```diff
@@ -161,6 +192,8 @@ handle_do(Scenario, UserIds, State) ->
 %% ------------------------------------------------------------------
 %% helpers
 %% ------------------------------------------------------------------
+
+-spec start_scenario(amoc:scenario(), [amoc_scenario:user_id()], state()) -> [term()].
```
Comment (gadgetci) [inline src/amoc_controller.erl]: According to **Elvis**:
> Line 196 is too long: -spec start_scenario(amoc:scenario(), [amoc_scenario:user_id()], state()) -> [term()]..

File: src/amoc_dist.erl [hunk 6]
```diff
@@ -11,41 +11,46 @@
...
 %% ------------------------------------------------------------------
 %% API
 %% ------------------------------------------------------------------
+-spec start_nodes() -> [ok].
 start_nodes() ->
     Hosts = application:get_env(amoc, hosts, []),
     Path = application:get_env(amoc, path, "/usr"),
     start_nodes(Hosts, Path).
 
+-spec do(amoc:scenario(), amoc_scenario:user_id(), amoc_scenario:user_id()) -> [any()].
```
Comment (gadgetci) [inline src/amoc_dist.erl]: According to **Elvis**:
> Line 23 is too long: -spec do(amoc:scenario(), amoc_scenario:user_id(), amoc_scenario:user_id()) -> [any()]..

File: src/amoc_controller.erl [hunk 7]
```diff
@@ -43,41 +48,55 @@
...
     gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
 
+-spec do(amoc:scenario(), amoc_scenario:user_id(), amoc_scenario:user_id()) ->
+    ok | {error, term()}.
 do(Scenario, Start, End) ->
     gen_server:call(?SERVER, {do, Scenario, Start, End}).
 
-do(Node, Scenario, Start, End, Nodes, NodeId, Opts) ->
-    gen_server:call({?SERVER, Node}, {do, Scenario, Start, End, Nodes, NodeId, Opts}).
+-spec do(node(), amoc:scenario(), amoc_scenario:user_id(), amoc_scenario:user_id(),
```
Comment (gadgetci) [inline src/amoc_controller.erl]: According to **Elvis**:
> Line 60 is too long: -spec do(node(), amoc:scenario(), amoc_scenario:user_id(), amoc_scenario:user_id(),.

File: src/amoc_dist.erl [hunk 8]
```diff
@@ -11,41 +11,46 @@
...
 do(Scenario, Start, End) ->
     do(Scenario, Start, End, []).
 
+-spec do(amoc:scenario(), amoc_scenario:user_id(), amoc_scenario:user_id(), amoc:do_opts()) -> [any()].
```
Comment (gadgetci) [inline src/amoc_dist.erl]: According to **Elvis**:
> Line 27 is too long: -spec do(amoc:scenario(), amoc_scenario:user_id(), amoc_scenario:user_id(), amoc:do_opts()) -> [any()]..

File: src/amoc_controller.erl [hunk 9]
```diff
@@ -43,41 +48,55 @@
...
+         non_neg_integer(), node_id(), amoc:do_opts()) -> ok | {error, term()}.
+do(Node, Scenario, Start, End, NodesCount, NodeId, Opts) ->
+    gen_server:call({?SERVER, Node}, {do, Scenario, Start, End, NodesCount, NodeId, Opts}).
```
Comment (gadgetci) [inline src/amoc_controller.erl]: According to **Elvis**:
> Line 63 is too long:     gen_server:call({?SERVER, Node}, {do, Scenario, Start, End, NodesCount, NodeId, Opts})..

File: src/amoc_metrics.erl [hunk 10]
```diff
@@ -34,18 +42,22 @@ new_counter(Name, Opts) when is_atom(Name) ->
...
 update_spiral(Name, Value) when is_atom(Name) ->
     ExName = [amoc, spirals, Name],
     exometer:update(ExName, Value).
 
+-spec update_counter(name(), integer()) -> ok.
 update_counter(Name, Value) when is_atom(Name) ->
     ExName = [amoc, counters, Name],
     exometer:update(ExName, Value).
 
+-spec reporter_and_interval([proplists:property()]) -> {module(), non_neg_integer()}.
```
Comment (gadgetci) [inline src/amoc_metrics.erl]: According to **Elvis**:
> Line 60 is too long: -spec reporter_and_interval([proplists:property()]) -> {module(), non_neg_integer()}..

File: src/amoc_controller.erl [hunk 11]
```diff
@@ -177,14 +211,18 @@ init_scenario(Scenario) ->
             skip
     end.
 
+-spec start_users(amoc:scenario(), [amoc_scenario:user_id()], state()) -> [term()].
```
Comment (gadgetci) [inline src/amoc_controller.erl]: According to **Elvis**:
> Line 214 is too long: -spec start_users(amoc:scenario(), [amoc_scenario:user_id()], state()) -> [term()]..

File: src/amoc_controller.erl [hunk 12]
```diff
@@ -193,19 +231,23 @@ stop_users(Users, _ForceRemove=true) ->
 stop_users(Users, _ForceRemove=false) ->
     [ Pid ! stop || {_, Pid} <- Users ].
 
+-spec last_users(non_neg_integer()) -> [{amoc_scenario:user_id(), pid()}].
 last_users(Count) ->
-    [ User || [User] <- last_users(Count, ets:last(?TABLE), []) ].
+    [ User || User <- last_users(Count, ets:last(?TABLE), []) ].
 
+-spec last_users(non_neg_integer(), amoc_scenario:user_id() | '$end_of_table',
+                 [{amoc_scenario:user_id(), pid()}]) -> [{amoc_scenario:user_id(), pid()}].
```
Comment (gadgetci) [inline src/amoc_controller.erl]: According to **Elvis**:
> Line 239 is too long:                  [{amoc_scenario:user_id(), pid()}]) -> [{amoc_scenario:user_id(), pid()}]..

File: src/amoc_slave.erl [hunk 13]
```diff
@@ -23,16 +23,22 @@
...
          code_change/3]).
 
--record(state, {to_ack, master}).
-
+-record(state, {to_ack :: [{node(), no_retries()}],
+                master :: node()}).
 -define(DEFAULT_RETRIES, 10).
 
+-type state() :: #state{}.
+-type command() :: {start, string(), file:filename()} | {monitor_master, node()}.
```
Comment (gadgetci) [inline src/amoc_slave.erl]: According to **Elvis**:
> Line 31 is too long: -type command() :: {start, string(), file:filename()} | {monitor_master, node()}..

File: src/amoc.erl [hunk 14]
```diff
@@ -3,16 +3,31 @@
...
 
+-export_type([
+              scenario/0,
+              do_opts/0,
+              remove_opts/0
+             ]).
+
+-type scenario() :: module().
+
+-type do_opt() :: {nodes, [node()]} | {comment, string()} | {repeat, timeout()} |
```
Comment (gadgetci) [inline src/amoc.erl]: According to **Elvis**:
> Line 20 is too long: -type do_opt() :: {nodes, [node()]} | {comment, string()} | {repeat, timeout()} |.

File: src/amoc_controller.erl [hunk 15]
```diff
@@ -135,15 +159,22 @@ handle_remove(Count, Opts, _State) when
...
 
+-spec handle_local_do(amoc:scenario(), amoc_scenario:user_id(), amoc_scenario:user_id(), state())->
+    {reply, ok | {error, term()}, state()}.
 handle_local_do(Scenario, Start, End, State) ->
     handle_do(Scenario, lists:seq(Start, End), State).
 
-handle_dist_do(Scenario, Start, End, Nodes, NodeId, Opts, State) ->
-    UserIds = node_userids(Start, End, Nodes, NodeId),
-    State1 = State#state{nodes = Nodes,
+-spec handle_dist_do(amoc:scenario(), amoc_scenario:user_id(), amoc_scenario:user_id(),
```
Comment (gadgetci) [inline src/amoc_controller.erl]: According to **Elvis**:
> Line 167 is too long: -spec handle_dist_do(amoc:scenario(), amoc_scenario:user_id(), amoc_scenario:user_id(),.

File: src/amoc_user.erl [hunk 16]
```diff
@@ -13,9 +13,13 @@
 -export([start_link/3]).
 -export([init/4]).
 
+-type state() :: term().
+
+-spec start_link(amoc:scenario(), amoc_scenario:user_id(), state()) -> {ok, pid()}.
 start_link(Scenario, Id, State) ->
     proc_lib:start_link(?MODULE, init, [self(), Scenario, Id, State]).
 
+-spec init(pid(), amoc:scenario(), amoc_scenario:user_id(), state()) -> no_return().
```
Comment (gadgetci) [inline src/amoc_user.erl]: According to **Elvis**:
> Line 22 is too long: -spec init(pid(), amoc:scenario(), amoc_scenario:user_id(), state()) -> no_return()..

File: src/amoc_user.erl [hunk 17]
```diff
@@ -13,9 +13,13 @@
 -export([start_link/3]).
 -export([init/4]).
 
+-type state() :: term().
+
+-spec start_link(amoc:scenario(), amoc_scenario:user_id(), state()) -> {ok, pid()}.
```
Comment (gadgetci) [inline src/amoc_user.erl]: According to **Elvis**:
> Line 18 is too long: -spec start_link(amoc:scenario(), amoc_scenario:user_id(), state()) -> {ok, pid()}..

File: src/amoc_controller.erl [hunk 18]
```diff
@@ -135,15 +159,22 @@ handle_remove(Count, Opts, _State) when
     Users = last_users(Count),
     stop_users(Users, ForceRemove).
 
+-spec handle_local_do(amoc:scenario(), amoc_scenario:user_id(), amoc_scenario:user_id(), state())->
```
Comment (gadgetci) [inline src/amoc_controller.erl]: According to **Elvis**:
> Line 162 is too long: -spec handle_local_do(amoc:scenario(), amoc_scenario:user_id(), amoc_scenario:user_id(), state())->.

File: src/amoc_annotations.erl [hunk 19]
```diff
@@ -17,12 +16,22 @@
 
 -record(state, {}).
 
+-type tag() :: binary().
+-type state() :: #state{}.
+
+-type command() :: {dist_do, amoc:scenario(), amoc_scenario:user_id(), amoc_scenario:user_id(), amoc:do_opts()} |
```
Comment (gadgetci) [inline src/amoc_annotations.erl]: According to **Elvis**:
> Line 22 is too long: -type command() :: {dist_do, amoc:scenario(), amoc_scenario:user_id(), amoc_scenario:user_id(), amoc:do_opts()} |.

**Comment by elbrujohalcon** [comment]
In case you're wondering @gadgetci's xref and compiler checks fail due to this error during compilation:
```bash
 In file included from c_src/exml_event.c:1:0:
c_src/exml_event.h:5:19: fatal error: expat.h: No such file or directory
 #include <expat.h>
                   ^
compilation terminated.
```

**Comment by ppikula** [comment]
@elbrujohalcon is is possible to install libexpat on the machine?

**Comment by elbrujohalcon** [comment]
@ppikula I think so… yes. Is that a requirement for this project to compile?

**Comment by elbrujohalcon** [comment]
@ignaciom, @demian711 : can you take care of that?

**Comment by ppikula** [comment]
@elbrujohalcon, @demian711 yes, it is required by the exml library. We use amoc for our xmpp load test, so It could be a little bit harder without xml parser ;). If you can, please install it. Thanks!

**Comment by elbrujohalcon** [comment]
@ppikula well… you can always use xmerl :trollface:

**Pull request merged**